NLP/main/extract_display_quotes.py

```diff
@@ -0,0 +1,531 @@
+                text_dict['text_id'] = doc_id
+                text_dict['text'] = doc_lines
+                all_files.append(text_dict)
+                    
+            except:
+                # this will provide some information in the case of an error
+                app_logger.exception("message")
+                traceback.print_exc()
+        
+        # convert the extracted texts into a pandas dataframe for further processing
+        text_df = pd.DataFrame.from_dict(all_files)
+        text_df['spacy_text'] = text_df['text'].apply(lambda text: self.nlp_preprocess(text))
+        text_df.set_index('text_id', inplace=True)
+        
+        return text_df
+    
+    
+    def process_xls(self, xls_upload):
+        '''
+        Pre-process uploaded .xlsx file into pandas dataframe
+
+        Args:
+            xls_upload: the uploaded .xlsx file from upload_files()
+        '''
+        # read the excel file containing the list of texts and convert them into a pandas dataframe
+        text_df = pd.read_excel(io.BytesIO(xls_upload.data[0]))
+        text_df['spacy_text'] = text_df['text'].apply(lambda text: self.nlp_preprocess(text))
+        text_df.set_index('text_id', inplace=True)
+        
+        return text_df
+        
+
+    def get_quotes(self, text_df, inc_ent, create_tree=False):
+        '''
+        Extract quotes and their meta-data (quote_id, quote_index, etc.) from the text
+        and return as a pandas dataframe
+
+        Args:
+            text_df: the pandas dataframe containing the list of texts
+            inc_ent: a list containing the named entities to be extracted from the text, 
+                     e.g., ['ORG','PERSON','GPE','NORP','FAC','LOC']
+            create_tree: option to create parse tree files for the quotes 
+        '''
+        # create an output folder and specify the file path if create_tree=True
+        if create_tree:
+            os.makedirs('output', exist_ok=True)
+            os.makedirs('output/trees', exist_ok=True)
+            tree_dir = './output/trees/'
+        else:
+            tree_dir = None
+    
+        # create an empty list to store all detected quotes
+        all_quotes = []
+        
+        # go through all the texts and start extracting quotes
+        for row in text_df.itertuples():
+            doc_id = row.Index
+            doc = row.spacy_text
+            
+            try:        
+                # extract the quotes
+                quotes = extract_quotes(doc_id=doc_id, doc=doc, 
+                                        write_tree=create_tree, 
+                                        tree_dir=tree_dir)
+                
+                # extract the included named entities
+                speaks, qts = [quote['speaker'] for quote in quotes], [quote['quote'] for quote in quotes]        
+                speak_ents = [[(str(ent), ent.label_) for ent in doc.ents if (str(ent) in speak) & (ent.label_ in inc_ent)] for speak in speaks]        
```

Comment:
This can also be generalised function for the same purpose, we may need to similar entity extraction from other types of string (other than speaker/quotation etc.)

Reply:
function created